api/app/Services/SingleSignOn/VanillaSingleSignOn.php

```diff
@@ -0,0 +1,297 @@
+<?php
+
+namespace App\Services\SingleSignOn;
+
+use Illuminate\Http\Request;
+
+class VanillaSingleSignOn extends SingleSignOnAbstract implements SingleSignOnContract
+{
+    /**
+     * Constant representing the max allowed timestamp difference in seconds.
+     *
+     * @var int
+     */
+    const TIMEOUT = 1440;
+
+    /**
+     * Client ID.
+     *
+     * @var string
+     */
+    protected $clientId;
+
+    /**
+     * Secret.
+     *
+     * @var string
+     */
+    protected $secret;
+
+    /**
+     * Security scheme to use.
+     *
+     * @var string|bool
+     */
+    protected $secure = 'sha1';
+
+    /**
+     * Assign dependencies.
+     *
+     * @param  Request  $request
+     * @param  mixed    $user
+     *
+     * @return void
+     */
+    public function __construct(Request $request, $user)
+    {
+        $this->clientId = env('VANILLA_JSCONNECT_CLIENT_ID');
+        $this->secret = env('VANILLA_JSCONNECT_SECRET');
+
+        parent::__construct($request, $user);
+    }
+
+    /**
+     * Get the response to the requester.
+     *
+     * @return mixed
+     */
+    public function getResponse()
+    {
+        return $this->response($this->formatUser(), $this->secure);
+    }
+
+    /**
+     * Format the user as expected by Vanilla.
+     *
+     * @return array
+     */
+    public function formatUser()
+    {
+        if (!$this->user) {
+            return [];
+        } else {
+            return [
+                'uniqueid' => $this->user->user_id,
+                'name' => $this->user->username,
+                'email' => $this->user->email,
+                'photourl' => $this->user->avatar_img_url,
+                'roles' => $this->getMappedRoles()
+            ];
+        }
+    }
+
+    /**
+     * Convert spira roles to a string suitable for Vanilla.
+     *
+     * @return string
+     */
+    protected function getMappedRoles()
+    {
+        $roles = $this->user->roles->pluck('name');
+
+        // Map Spira to Vanilla roles
+        $roles = $roles->map(function ($role) {
+            $mapping = [
+                'admin' => 'administrator',
+                'guest' => 'member'
+            ];
+
+            if (array_key_exists($role, $mapping)) {
+                return $mapping[$role];
+            }
+
+            return $role;
+        });
+
+        return $roles->implode(',');
+    }
+
+    /**
+     * Generate the response string that Vanilla expects.
+     *
+     * @param  array  $user
+     * @param  mixed  $secure
+     *
+     * @return string
+     */
+    protected function response($user, $secure = true)
```

Comment:
@artstorm ah! that makes more sense. Yes, especially as this code is within our /api , I would like it not to be so horrible. I'm glad that wasn't your work, you had me worried!